File: examples/python/Vpp/virtual_patern_projection.py
import cv2 as cv
import depthai as dai

"""
  ┌──────┐       ┌───────────────┐ ----------------left------------------------------> ┌─────┐
  | Left | ----> |               |                                                     |     |
  └──────┘       |               | ----------------right-----------------------------> |     | -left--> ┌────────┐
                 | Rectification |                                                     | Vpp |          | Stereo | --depth->
  ┌───────┐      |               | --left_low_res---> ┌──────────────┐ --disparity---> |     |          |        |
  | Right | ---> |               |                    | NeuralStereo |                 |     | -right-> └────────┘
  └───────┘      └───────────────┘ --right_low_res--> └──────────────┘ --confidence--> └─────┘
"""


if __name__ == "__main__":
    fps = 20

    device = dai.Device()
    pipeline = dai.Pipeline(device)

    if not device.isNeuralDepthSupported():
        print("Exiting Vpp example: device doesn't support NeuralDepth.")
        exit()

    # Left Right cameras
    monoLeft = pipeline.create(dai.node.Camera).build(dai.CameraBoardSocket.CAM_B, sensorFps=fps)
    monoRight = pipeline.create(dai.node.Camera).build(dai.CameraBoardSocket.CAM_C, sensorFps=fps)
    monoLeftOut = monoLeft.requestFullResolutionOutput()
    monoRightOut = monoRight.requestFullResolutionOutput()

    # Rectification node
    rectification = pipeline.create(dai.node.Rectification)
    monoLeftOut.link(rectification.input1)
    monoRightOut.link(rectification.input2)

    # Neural Depth
    neuralDepth = pipeline.create(dai.node.NeuralDepth).build(
        monoLeftOut,
        monoRightOut,
        dai.DeviceModelZoo.NEURAL_DEPTH_NANO
    )

    # Vpp node
    vpp = pipeline.create(dai.node.Vpp).build(
        rectification.output1,
        rectification.output2,
        neuralDepth.disparity,
        neuralDepth.confidence
    )
    # set initial parameters
    vpp.initialConfig.blending = 0.5
    vpp.initialConfig.maxPatchSize = 2
    vpp.initialConfig.patchColoringType = dai.VppConfig.PatchColoringType.RANDOM
    vpp.initialConfig.uniformPatch = True
    vpp.initialConfig.maxFPS = fps
    injection_params = vpp.initialConfig.injectionParameters
    injection_params.textureThreshold = 3.0
    injection_params.useInjection = True

    # for runtime parameter control use vpp.inputConfig queue
    # config_queue = vpp.inputConfig.createInputQueue()
    # config_queue.send(vpp_config)

    # Stereo node
    stereo_node = pipeline.create(dai.node.StereoDepth)
    stereo_node.setRectification(False)
    vpp.leftOut.link(stereo_node.left)
    vpp.rightOut.link(stereo_node.right)

    # Create output queues we want to show
    vpp_output_left = vpp.leftOut.createOutputQueue()
    vpp_output_right = vpp.rightOut.createOutputQueue()
    depth_queue = stereo_node.depth.createOutputQueue()

    # Start the pipeline
    with pipeline:
        pipeline.start()
        print("Pipeline started successfully")
        while pipeline.isRunning():
            vpp_out_left = vpp_output_left.get()
            vpp_out_right = vpp_output_right.get()
            depth = depth_queue.get()

            cv.imshow("vpp_left", vpp_out_left.getCvFrame())
            cv.imshow("vpp_right", vpp_out_right.getCvFrame())
            cv.imshow("Depth", dai.colorizeDepthFrame(depth, 500, 12000, cv.COLORMAP_TURBO, False).getCvFrame())

            key = cv.waitKey(1)
            if key == ord('q'):
                quit()

File: examples/cpp/Vpp/virtual_patern_projection.cpp
#include <depthai/depthai.hpp>
#include <iostream>
#include <opencv2/opencv.hpp>

/**
  ┌──────┐       ┌───────────────┐ ----------------left------------------------------> ┌─────┐
  | Left | ----> |               |                                                     |     |
  └──────┘       |               | ----------------right-----------------------------> |     | -left--> ┌────────┐
                 | Rectification |                                                     | Vpp |          | Stereo | --depth->
  ┌───────┐      |               | --left_low_res---> ┌──────────────┐ --disparity---> |     |          |        |
  | Right | ---> |               |                    | NeuralStereo |                 |     | -right-> └────────┘
  └───────┘      └───────────────┘ --right_low_res--> └──────────────┘ --confidence--> └─────┘
**/

int main() {
    int fps = 20;
    dai::Pipeline pipeline;
    auto device = pipeline.getDefaultDevice();
    if(!device->isNeuralDepthSupported()) {
        std::cout << "Exiting Vpp example: device doesn't support NeuralDepth.\n";
        return 0;
    }

    // Left / Right cameras
    auto monoLeft = pipeline.create<dai::node::Camera>()->build(dai::CameraBoardSocket::CAM_B, std::nullopt, fps);
    auto monoRight = pipeline.create<dai::node::Camera>()->build(dai::CameraBoardSocket::CAM_C, std::nullopt, fps);

    auto monoLeftOut = monoLeft->requestFullResolutionOutput();
    auto monoRightOut = monoRight->requestFullResolutionOutput();

    // Rectification node
    auto rectification = pipeline.create<dai::node::Rectification>();
    monoLeftOut->link(rectification->input1);
    monoRightOut->link(rectification->input2);

    // Neural Depth builder
    auto neuralDepth = pipeline.create<dai::node::NeuralDepth>()->build(*monoLeftOut, *monoRightOut, dai::DeviceModelZoo::NEURAL_DEPTH_NANO);

    // VPP node builder
    auto vpp = pipeline.create<dai::node::Vpp>()->build(rectification->output1, rectification->output2, neuralDepth->disparity, neuralDepth->confidence);

    // Set initial parameters
    vpp->initialConfig->blending = 0.5f;
    vpp->initialConfig->maxPatchSize = 2;
    vpp->initialConfig->patchColoringType = dai::VppConfig::PatchColoringType::RANDOM;
    vpp->initialConfig->uniformPatch = true;
    vpp->initialConfig->maxFPS = static_cast<float>(fps);

    auto& injectionParameters = vpp->initialConfig->injectionParameters;
    injectionParameters.textureThreshold = 3.0f;
    injectionParameters.useInjection = true;

    // Stereo depth node
    auto stereoNode = pipeline.create<dai::node::StereoDepth>();
    stereoNode->setRectification(false);

    vpp->leftOut.link(stereoNode->left);
    vpp->rightOut.link(stereoNode->right);

    // Output queues
    auto vppLeftQueue = vpp->leftOut.createOutputQueue();
    auto vppRightQueue = vpp->rightOut.createOutputQueue();
    auto depthQueue = stereoNode->depth.createOutputQueue();

    std::cout << "Pipeline started successfully" << std::endl;

    pipeline.start();

    while(true) {
        auto vppLeftFrame = vppLeftQueue->get<dai::ImgFrame>();
        auto vppRightFrame = vppRightQueue->get<dai::ImgFrame>();
        auto depthFrame = depthQueue->get<dai::ImgFrame>();

        cv::imshow("vppLeft", vppLeftFrame->getCvFrame());
        cv::imshow("vppRight", vppRightFrame->getCvFrame());

        cv::imshow("Depth", dai::utility::colorizeDepthFrame(*depthFrame, 500.0f, 12000.0f, cv::COLORMAP_TURBO, true).getCvFrame());

        if(cv::waitKey(1) == 'q') {
            break;
        }
    }

    return 0;
}
